### PR TITLE
Radix 4 Division by digit recurrence

### PR DESCRIPTION
**Pull request by Durchbruchswagen**
Radix-4 division by digit-recurrence algorithm (precisely its SRT variant). This algorithm works on floating-point numbers that are greater than or equal to `1/2` and less than `1` (basically, the normalization bit is at $2^{−1}$ rather than at $2^0$, as is the case for IEEE floats). 

The implementation is based on `Digital Arithmetic` by `Miloš D. Ercegovac` and `Tomás Lang`.

TODO:

- [x]  Add more comments to explain the algorithm
- [x] Add more tests from the book
- [x] Add test that gonna result in negative residual to test the quotient correction
- [x] Refactor parts of code
- [x] Change the body of division to FSM

### CODE REVIEW

File: coreblocks/func_blocks/fu/fpu/dr_division.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+                with m.If(self.div_result.run):
+                    m.d.sync += result_ready.eq(0)
+                    m.d.sync += init_ready.eq(1)
+                    m.next = "Idle"
+
+        @def_method(m, self.div_result, ready=result_ready)
+        def _():
+            zero_rem = Signal()
+            adjusted_result = Signal(self.div_params.result_fractional_bits)
+            # This is note for the future. We are chcecking if the residual
```
Comment (qbojj) [inline coreblocks/func_blocks/fu/fpu/dr_division.py]: maybe mark this with an explicit TODO tag for better greppability

File: coreblocks/func_blocks/fu/fpu/dr_division.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        result_ready = Signal()
+        residual_negative = Signal()
+        residual_is_zero = Signal()
+        residual_is_minus_d = Signal()
+        otfc_response = Signal(from_method_layout(otfc.method_layouts.otfc_result_out_layout))
+        qsf_response = Signal(from_method_layout(qsf.method_layouts.qsf_out_layout))
+
+        @def_method(m, self.div_init, ready=init_ready)
+        def _(x, d):
+            m.d.sync
```
Comment (qbojj) [inline coreblocks/func_blocks/fu/fpu/dr_division.py]: ?

File: coreblocks/func_blocks/fu/fpu/dr_division.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            m.d.sync += counter.eq(0)
+            m.d.sync += result_ready.eq(0)
+
+        with m.FSM(init="Idle"):
+            with m.State("Idle"):
+                with m.If(self.div_init.run):
+                    m.next = "Loop"
+                    m.d.sync += init_ready.eq(0)
+            with m.State("Loop"):
+                with m.If((counter < (counter_max))):
```
Comment (qbojj) [inline coreblocks/func_blocks/fu/fpu/dr_division.py]:
```py
with m.If(counter < counter_max):
    ...
with m.Else():
    ...
```
or even skip the `If` and just move to the correct next state directly.

File: coreblocks/func_blocks/fu/fpu/dr_division.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                        )
+                        m.d.comb += qsf_response.eq(resp_qsf)
+                        otfc.otfc_add_digit(m, sign=qsf_response["sign"], q=qsf_response["q"])
+                    m.d.sync += counter.eq(counter + 1)
+                    # To check if the last residual is zero we keep this
+                    # information in a separate flag before we compute new residual
+                    # This applies to the other flags as well
+                    m.d.sync += residual_negative.eq(residual < 0)
+                    m.d.sync += residual_is_minus_d.eq(residual == m_one_p)
+                    m.d.sync += residual_is_zero.eq(~(residual.any()))
```
Comment (qbojj) [inline coreblocks/func_blocks/fu/fpu/dr_division.py]: I think more explicit `residual == 0` is clearer

File: coreblocks/func_blocks/fu/fpu/dr_division.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                    new_residual = Signal(self.div_params.fractional_bits + 1)
+
+                    with Transaction().body(m):
+                        resp_qsf = qsf.qsf_request(
+                            m,
+                            residual=residual[-7:].as_signed(),
+                            divisor=(divisor[-5:] << 0),
+                        )
+                        m.d.comb += qsf_response.eq(resp_qsf)
+                        otfc.otfc_add_digit(m, sign=qsf_response["sign"], q=qsf_response["q"])
```
Comment (qbojj) [inline coreblocks/func_blocks/fu/fpu/dr_division.py]:
```suggestion
                        m.d.av_comb += qsf_response.eq(qsf.qsf_request(
                            m,
                            residual=residual[-7:].as_signed(),
                            divisor=divisor[-5:],
                        ))
                        otfc.otfc_add_digit(m, sign=qsf_response["sign"], q=qsf_response["q"])
```
should probably also have an assertion (kuznia-rdzeni/transactron#198)

File: coreblocks/func_blocks/fu/fpu/dr_division.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+        init_ready = Signal(init=1)
+        result_ready = Signal()
+        residual_negative = Signal()
+        residual_is_zero = Signal()
+        residual_is_minus_d = Signal()
+        otfc_response = Signal(from_method_layout(otfc.method_layouts.otfc_result_out_layout))
+        qsf_response = Signal(from_method_layout(qsf.method_layouts.qsf_out_layout))
+
+        @def_method(m, self.div_init, ready=init_ready)
```
Comment (qbojj) [inline coreblocks/func_blocks/fu/fpu/dr_division.py]: `init_ready` and `result_ready` could be done using `fsm.ongoing(STATE)`, or the bodies of the method could be put inside the FSM allowing for simple `m.next` inside of the body

File: coreblocks/func_blocks/fu/fpu/dr_division.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+                resp = otfc.otfc_result(m, shift=0)
+                m.d.comb += otfc_response.eq(resp)
+            # The initialization condition requires that we shift result left by 2.
+            # We can also do this by setting correct signal shape
+            with m.If(residual_negative):
+                m.d.comb += zero_rem.eq(residual_is_minus_d)
+                m.d.comb += adjusted_result.eq((otfc_response["result"] - 1))
+            with m.Else():
+                m.d.comb += zero_rem.eq(residual_is_zero)
+                m.d.comb += adjusted_result.eq(otfc_response["result"])
```
Comment (qbojj) [inline coreblocks/func_blocks/fu/fpu/dr_division.py]: use `m.d.av_comb`

File: coreblocks/func_blocks/fu/fpu/dr_division.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+            adjusted_result = Signal(self.div_params.result_fractional_bits)
+            # This is note for the future. We are chcecking if the residual
+            # is not equal -d beacuse in this case of subtracting 1 from the result
+            # to correct the negative final residual, our residual would be zero
+            # However the books doesn't mention that something like that is needed
+            # It just checks if the last residual is zero, so maybe this is redundant
+            m.d.sync += counter.eq(0)
+            with Transaction().body(m):
+                resp = otfc.otfc_result(m, shift=0)
+                m.d.comb += otfc_response.eq(resp)
```
Comment (qbojj) [inline coreblocks/func_blocks/fu/fpu/dr_division.py]:
```suggestion
            m.d.av_comb += otfc_response.eq(otfc.otfc_result(m, shift=0))
```

File: test/func_blocks/fu/fpu/test_dr_division.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+            iterations=7,
+            fractional_bits=10,
+            result_fractional_bits=14,
+        )
+        drd = SimpleTestCircuit(DrDivModule(div_params=params, qsf_params=R4A2RED_PARAMS))
+
+        async def tests(sim: TestbenchContext):
+            input_dict = {}
+            test_cases = [
+                TCase(2**9, 2**9, int("1000000000000", 2), 1),
```
Comment (qbojj) [inline test/func_blocks/fu/fpu/test_dr_division.py]: just use `0b*` literals

**Review by qbojj — CHANGES_REQUESTED** [review]

File: coreblocks/func_blocks/fu/fpu/dr_division.py (hunk 1 above)
Comment (Durchbruchswagen) [reply coreblocks/func_blocks/fu/fpu/dr_division.py]: I thought about it, and this case is not necessary. The algorithm returns results in the form $d*q ± r$. The $-$ is used in case we overshoot due to the redundant digit set, but $r$ is always less than $d$, so this is most likely why the book omitted this case. I have removed this comment altogether.

File: coreblocks/func_blocks/fu/fpu/dr_division.py (hunk 2 above)
Comment (Durchbruchswagen) [reply coreblocks/func_blocks/fu/fpu/dr_division.py]: Sorry, that was a leftover from a code refactoring. I have removed the line.

File: coreblocks/func_blocks/fu/fpu/dr_division.py (hunk 3 above)
Comment (Durchbruchswagen) [reply coreblocks/func_blocks/fu/fpu/dr_division.py]: I removed the `if/elif` block and changed it to a simple `if` statement that moves to the next state on the last cycle.

File: coreblocks/func_blocks/fu/fpu/dr_division.py (hunk 4 above)
Comment (Durchbruchswagen) [reply coreblocks/func_blocks/fu/fpu/dr_division.py]: Done

File: coreblocks/func_blocks/fu/fpu/dr_division.py (hunk 5 above)
Comment (Durchbruchswagen) [reply coreblocks/func_blocks/fu/fpu/dr_division.py]: Done! I removed the intermediate signal and used `always_body` in the transaction.

File: coreblocks/func_blocks/fu/fpu/dr_division.py (hunk 6 above)
Comment (Durchbruchswagen) [reply coreblocks/func_blocks/fu/fpu/dr_division.py]: Done. The methods for `init` and `result` are now moved inside the FSM states and use `ongoing` for the `ready` signal.

File: coreblocks/func_blocks/fu/fpu/dr_division.py (hunk 7 above)
Comment (Durchbruchswagen) [reply coreblocks/func_blocks/fu/fpu/dr_division.py]: Done

File: coreblocks/func_blocks/fu/fpu/dr_division.py (hunk 8 above)
Comment (Durchbruchswagen) [reply coreblocks/func_blocks/fu/fpu/dr_division.py]: Done

File: test/func_blocks/fu/fpu/test_dr_division.py (hunk 9 above)
Comment (Durchbruchswagen) [reply test/func_blocks/fu/fpu/test_dr_division.py]: Done

File: coreblocks/func_blocks/fu/fpu/dr_division.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+        residual_negative = Signal()
+        residual_is_zero = Signal()
+        residual_is_minus_d = Signal()
+        otfc_response = Signal(from_method_layout(otfc.method_layouts.otfc_result_out_layout))
+        qsf_response = Signal(from_method_layout(qsf.method_layouts.qsf_out_layout))
+
+        with m.FSM(init="Idle") as fsm:
+            with m.State("Idle"):
+
+                @def_method(m, self.div_init, ready=fsm.ongoing("Idle"))
```
Comment (qbojj) [inline coreblocks/func_blocks/fu/fpu/dr_division.py:118]: now that the body is in the FSM, there is no need for the explicit `ready`.

File: coreblocks/func_blocks/fu/fpu/dr_division.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+                            m.d.comb += new_residual.eq(residual - one_p)
+                    with m.Case(0):
+                        m.d.comb += new_residual.eq(residual)
+                m.d.sync += residual.eq(new_residual << 2)  # R[j + 1] = 4*R[j]
+                next_counter = counter + 1
+                with m.If(next_counter == counter_max):
+                    m.next = "Result"
+            with m.State("Result"):
+
+                @def_method(m, self.div_result, ready=fsm.ongoing("Result"))
```
Comment (qbojj) [inline coreblocks/func_blocks/fu/fpu/dr_division.py:178]: same here

**Review by qbojj — APPROVED** [review]
just remove the ready signals and we can merge

**Review by qbojj — APPROVED** [review]

File: coreblocks/func_blocks/fu/fpu/dr_division.py (hunk 11 above)
Comment (Durchbruchswagen) [reply coreblocks/func_blocks/fu/fpu/dr_division.py]: Done

File: coreblocks/func_blocks/fu/fpu/dr_division.py (hunk 10 above)
Comment (Durchbruchswagen) [reply coreblocks/func_blocks/fu/fpu/dr_division.py]: Done
